service/collector/bridge.go

```diff
@@ -0,0 +1,164 @@
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"regexp"
+
+	"github.com/giantswarm/apiextensions/pkg/clientset/versioned"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"github.com/prometheus/client_golang/prometheus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	subsystemBridge = "bridge"
+)
+
+var (
+	bridgeClusterWithoutNetworkInterfaceDesc *prometheus.Desc = prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, subsystemBridge, "cluster_without_network_interface"),
+		"Clusters without network interface.",
+		[]string{
+			labelCluster,
+		},
+		nil,
+	)
+	bridgeNetworkInterfaceWithoutClusterDesc *prometheus.Desc = prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, subsystemBridge, "network_interface_without_cluster"),
+		"Network interfaces without cluster.",
+		[]string{
+			labelCluster,
+		},
+		nil,
+	)
+)
+
+type BridgeConfig struct {
+	G8sClient versioned.Interface
+	Logger    micrologger.Logger
+}
+
+type Bridge struct {
+	g8sClient versioned.Interface
+	logger    micrologger.Logger
+}
+
+func NewBridge(config BridgeConfig) (*Bridge, error) {
+	if config.G8sClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.G8sClient must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	c := &Bridge{
+		g8sClient: config.G8sClient,
+		logger:    config.Logger,
+	}
+
+	return c, nil
+}
+
+func (c *Bridge) Collect(ch chan<- prometheus.Metric) error {
+	ctx := context.Background()
+
+	var desiredClusterIDs []string
+	{
+		c.logger.LogCtx(ctx, "level", "debug", "message", "finding all cluster IDs in FlannelConfigs")
+
+		l, err := c.g8sClient.Core().FlannelConfigs(metav1.NamespaceAll).List(metav1.ListOptions{})
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		for _, i := range l.Items {
+			desiredClusterIDs = append(desiredClusterIDs, i.Name)
+		}
+
+		c.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("found %d cluster IDs: %#v", len(desiredClusterIDs), desiredClusterIDs))
+	}
+
+	var bridgeClusterIDs []string
+	{
+		c.logger.LogCtx(ctx, "level", "debug", "message", "finding network interfaces")
+
+		interfaces, err := net.Interfaces()
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		for _, i := range interfaces {
+			id := clusterIDFromName(i.Name)
+			if id == "" {
+				// There are many different network interfaces and we cannot parse all
+				// of them. Thus we continue and go ahead with the next one we found.
+				continue
+			}
+
+			bridgeClusterIDs = append(bridgeClusterIDs, id)
+		}
+
+		c.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("found %d network interfaces for cluster IDs: %#v", len(bridgeClusterIDs), bridgeClusterIDs))
+	}
+
+	{
+		l, r := symmetricDifference(desiredClusterIDs, bridgeClusterIDs)
+
+		if len(l) == 0 {
+			c.logger.LogCtx(ctx, "level", "debug", "message", "no orphaned cluster IDs found")
+		}
```

Comment:
When success case is logged, I think it would be very important to also see number (and values) of leftover IDs in the logs as well.

Reply:
That means it is in metrics and logs. You want to have both? I would then again provide a separate PR to do this for both collectors as they are more or less identical in their behaviour.

Reply:
Metrics are useful to see trend over time. Logs are useful to see present live situation when shit has hit the fan. When on-call person is woken up at 3 am or working under pressure when production is on fire, having these kind of things explicit helps. That's why I'd like to have both.

Reply:
Makes sense. Approve and I provide pull requests to address your concerns separately, also for the other collector.

Reply:
We have a deal.